warehouse/manage/views.py

```diff
@@ -1059,6 +1069,206 @@ def toggle_2fa_requirement(self):
+    has_translations=True,
+    require_reauth=True,
+    http_cache=0,
+)
+class ManageOIDCProviderViews:
+    def __init__(self, project, request):
+        self.request = request
+        self.project = project
+
+    @property
+    def _ratelimiters(self):
+        return {
+            "user.oidc": self.request.find_service(
+                IRateLimiter, name="user_oidc.provider.register"
+            ),
+            "ip.oidc": self.request.find_service(
+                IRateLimiter, name="ip_oidc.provider.register"
+            ),
+        }
+
+    def _hit_ratelimits(self):
+        self._ratelimiters["user.oidc"].hit(self.request.user.id)
+        self._ratelimiters["ip.oidc"].hit(self.request.remote_addr)
+
+    def _check_ratelimits(self):
+        if not self._ratelimiters["user.oidc"].test(self.request.user.id):
+            raise TooManyOIDCRegistrations(
+                resets_in=self._ratelimiters["user.oidc"].resets_in(
+                    self.request.user.id
+                )
+            )
+
+        if not self._ratelimiters["ip.oidc"].test(self.request.remote_addr):
+            raise TooManyOIDCRegistrations(
+                resets_in=self._ratelimiters["ip.oidc"].resets_in(self.remote_addr)
+            )
+
+    @property
+    def default_response(self):
+        return {
+            "project": self.project,
+            "github_provider_form": GitHubProviderForm(
+                api_token=self.request.registry.settings.get("github.token"),
+            ),
+        }
+
+    @view_config(request_method="GET")
+    def manage_project_oidc_providers(self):
+        if self.request.flags.enabled(AdminFlagValue.DISALLOW_OIDC):
+            self.request.session.flash(
+                (
+                    "OpenID Connect is temporarily disabled. "
+                    "See https://pypi.org/help#admin-intervention for details."
+                ),
+                queue="error",
+            )
+
+        return self.default_response
+
+    @view_config(request_method="POST", request_param=GitHubProviderForm.__params__)
+    def add_github_oidc_provider(self):
+        if self.request.flags.enabled(AdminFlagValue.DISALLOW_OIDC):
+            self.request.session.flash(
+                (
+                    "OpenID Connect is temporarily disabled. "
+                    "See https://pypi.org/help#admin-intervention for details."
+                ),
+                queue="error",
+            )
+            return self.default_response
+
+        try:
+            self._check_ratelimits()
+        except TooManyOIDCRegistrations as exc:
+            return HTTPTooManyRequests(
+                self.request._(
+                    "There have been too many attempted OpenID Connect registrations. "
+                    "Try again later."
+                ),
+                retry_after=exc.resets_in.total_seconds(),
+            )
+
+        form = GitHubProviderForm(
+            self.request.POST,
```

Comment:
Is there a best practice for this documented anywhere (for either Pyramid or Warehouse in particular)? Some of the other views use the bare `self.request.POST`, e.g.:

```python
    @view_config(
        request_method="POST",
        request_param=AddEmailForm.__params__,
        require_reauth=True,
    )
    def add_email(self):
        form = AddEmailForm(
            self.request.POST,
            user_service=self.user_service,
            user_id=self.request.user.id,
        )
        
        # ...
```

I vaguely remember someone telling me *not* to use the splat form because it could conceivably allow the other (trusted) kwargs to be injected, like `user_id` in the example above. But I could be misremembering that.

Reply:
Yeah, the WTForms documentation supports (and gives priority to) passing the `dict` in directly, rather than splatting it: https://wtforms.readthedocs.io/en/3.0.x/forms/#wtforms.form.Form.__init__

Specifically:

> kwargs – Merged with data to allow passing existing data as parameters. Overwrites any duplicate keys in data. Only used if formdata is not passed.

Reply:
Yeah, great point and TIL. Thanks for clarifying.